// @amplitude/analytics-core/lib/esm/index.js
export { AmplitudeCore } from './core-client';
export { Identify } from './identify';
export { Revenue, RevenueProperty } from './revenue';
export { Destination } from './plugins/destination';
export { IdentityEventSender } from './plugins/identity';
export { Config, RequestMetadata } from './config';
export { Logger } from './logger';
export { getGlobalScope } from './global-scope';
export { getAnalyticsConnector, setConnectorDeviceId, setConnectorUserId } from './analytics-connector';
export { isNewSession } from './session';
export { getCookieName, getOldCookieName } from './cookie-name';
export { getLanguage } from './language';
export { getQueryParams } from './query-params';
export { returnWrapper } from './utils/return-wrapper';
export { debugWrapper, getClientLogConfig, getClientStates } from './utils/debug';
export { UUID } from './utils/uuid';
export { createIdentifyEvent } from './utils/event-builder';
export { isUrlMatchAllowlist, getDecodeURI } from './utils/url-utils';
export { generateHashCode, isTimestampInSample } from './utils/sampling';
export { MemoryStorage } from './storage/memory';
export { CookieStorage } from './storage/cookie';
export { getStorageKey } from './storage/helpers';
export { BrowserStorage } from './storage/browser-storage';
export { DiagnosticsClient } from './diagnostics/diagnostics-client';
export { registerSdkLoaderMetadata } from './diagnostics/uncaught-sdk-errors';
export { BaseTransport } from './transports/base';
export { FetchTransport } from './transports/fetch';
export { RemoteConfigClient } from './remote-config/remote-config';
export { LogLevel } from './types/loglevel';
export { AMPLITUDE_PREFIX, STORAGE_PREFIX } from './types/constants';
export { IdentifyOperation, SpecialEventType, } from './types/event/event';
export { ServerZone } from './types/server-zone';
export { OfflineDisabled } from './types/offline';
export { DEFAULT_CSS_SELECTOR_ALLOWLIST, DEFAULT_DATA_ATTRIBUTE_PREFIX, DEFAULT_ACTION_CLICK_ALLOWLIST, } from './types/element-interactions';
export { DEFAULT_DEAD_CLICK_ALLOWLIST, DEFAULT_RAGE_CLICK_ALLOWLIST, DEFAULT_RAGE_CLICK_THRESHOLD, DEFAULT_RAGE_CLICK_WINDOW_MS, DEFAULT_RAGE_CLICK_OUT_OF_BOUNDS_THRESHOLD, DEFAULT_DEAD_CLICK_WINDOW_MS, } from './types/frustration-interactions';
export { Status } from './types/status';
export { NetworkEventCallback, networkObserver } from './network-observer';
export { NetworkRequestEvent } from './network-request-event';
export { SAFE_HEADERS, FORBIDDEN_HEADERS } from './types/constants';
export { EMPTY_VALUE, BASE_CAMPAIGN, MKTG } from './types/constants';
export { CampaignParser } from './campaign/campaign-parser';
export { getPageTitle, TEXT_MASK_ATTRIBUTE, MASKED_TEXT_VALUE, replaceSensitiveString, CC_REGEX, SSN_REGEX, EMAIL_REGEX, } from './plugins/helpers';
export { Observable, asyncMap, merge, multicast } from './utils/observable';
